server/src/main/java/org/apache/iotdb/db/mpp/buffer/DataBlockManager.java

```diff
@@ -19,98 +19,296 @@
-    private String instanceId;
-
-    public FragmentInstanceInfo(
-        String hostname, String queryId, String fragmentId, String instanceId) {
-      this.hostname = Validate.notNull(hostname);
-      this.queryId = Validate.notNull(queryId);
-      this.fragmentId = Validate.notNull(fragmentId);
-      this.instanceId = Validate.notNull(instanceId);
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+
+public class DataBlockManager implements IDataBlockManager {
+
+  private static final Logger logger = LoggerFactory.getLogger(DataBlockManager.class);
+
+  public interface SourceHandleListener {
+    void onFinished(SourceHandle sourceHandle);
+
+    void onClosed(SourceHandle sourceHandle);
+  }
+
+  public interface SinkHandleListener {
+    void onFinish(SinkHandle sinkHandle);
+
+    void onClosed(SinkHandle sinkHandle);
+
+    void onAborted(SinkHandle sinkHandle);
+  }
+
+  /** Handle thrift communications. */
+  class DataBlockServiceImpl implements DataBlockService.Iface {
+
+    public GetDataBlockResponse getDataBlock(GetDataBlockRequest req) throws TException {
+      logger.debug(
+          "Get data block request received. Asking for data block [{}, {}) from {}.",
+          req.getStartSequenceId(),
+          req.getEndSequenceId(),
+          req.getSourceFragnemtInstanceId());
+      if (!sinkHandles.containsKey(req.getSourceFragnemtInstanceId())) {
+        throw new TException(
+            "Source fragment instance not found. Fragment instance ID: "
+                + req.getSourceFragnemtInstanceId()
+                + ".");
+      }
+      GetDataBlockResponse resp = new GetDataBlockResponse();
+      for (int i = req.getStartSequenceId(); i < req.getEndSequenceId(); i++) {
+        ByteBuffer serializedTsBlock =
+            sinkHandles
+                .get(req.getSourceFragnemtInstanceId())
+                .getSerializedTsBlock(req.getStartSequenceId());
+        resp.addToTsBlocks(serializedTsBlock);
+      }
+      return resp;
     }
 
-    public String getHostname() {
-      return hostname;
+    @Override
+    public void onNewDataBlockEvent(NewDataBlockEvent e) throws TException {
+      logger.debug(
+          "New data block event received, for operator {} of {} from {}.",
+          e.getTargetOperatorId(),
+          e.getTargetFragmentInstanceId(),
+          e.getSourceFragnemtInstanceId());
+      if (!sourceHandles.containsKey(e.getTargetFragmentInstanceId())
+          || !sourceHandles
+              .get(e.getTargetFragmentInstanceId())
+              .containsKey(e.getTargetOperatorId())
+          || sourceHandles
+              .get(e.getTargetFragmentInstanceId())
+              .get(e.getTargetOperatorId())
+              .isClosed()) {
+        logger.info("{} is ignored because the source handle does not exist or has been closed", e);
+        return;
```

Comment:
If the NewDataBlockEvent is omitted because the corresponding SourceHandle is not created yet, the SourceHandle will keep the `blocked` status after being creating in the future until next event comes.

And if the SinkHandle cannot send the next event due to memory limit or there is no more TsBlocks to be sent, the SourceHandle may be blocked permanently...

Maybe we can create the SourceHandle and put it in the map if it hasn't be created when receiving the NewDataBlockEvent

Reply:
Good catch. I forget to change this after I changed the transmission protocol.

> If the NewDataBlockEvent is omitted because the corresponding SourceHandle is not created yet, the SourceHandle will keep the blocked status after being creating in the future until next event comes.
True. Actually, it is not likely to happen since the series scan operator takes some time to produce data for downstream instances, I believe instances should be available then.

> And if the SinkHandle cannot send the next event due to memory limit or there is no more TsBlocks to be sent, the SourceHandle may be blocked permanently...
The SinkHandle will attempt to send the events at most 3 times. If it doesn't work, the SinkHandle throws an 
I/O exception on next invocation of any method, telling the caller that there is something wrong.